Hi @cjolivier01, @kpatelatwork,

The change looks good to me.

However, there are tests failures:
```
[ERROR] Failures: 
[ERROR]   S3OutputStreamTest.testPropagateOtherRetriableS3Exceptions:50 Multipart upload failed to complete.: unexpected exception type thrown; expected:<org.apache.kafka.connect.errors.ConnectException> but was:<java.io.IOException>
[ERROR]   S3OutputStreamTest.testPropagateRetriableS3Exceptions:44 Multipart upload failed to complete.: unexpected exception type thrown; expected:<org.apache.kafka.connect.errors.ConnectException> but was:<java.io.IOException>
[ERROR]   S3OutputStreamTest.testPropagateUnretriableS3Exceptions:35 Unable to initiate Multipart Upload.: unexpected exception type thrown; expected:<org.apache.kafka.connect.errors.ConnectException> but was:<java.io.IOException>
[INFO] 
[ERROR] Tests run: 258, Failures: 3, Errors: 0, Skipped: 0
```

@cjolivier01, could you please fix it?

--- a/kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java
+++ b/kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java
@@ -163,9 +163,21 @@ public class S3OutputStream extends PositionOutputStream {
       multiPartUpload.complete();
       log.debug("Upload complete for bucket '{}' key '{}'", bucket, key);
     } catch (IOException e) {
-      // Catch here in order to log this specific error and then rethrow.
-      log.error("Multipart upload failed to complete for bucket '{}' key '{}'", bucket, key);
-      throw e;
+      // Catch here in order to log this specific error and then rethrow
+      // with a more specific message (message may be used for error mapping).
+      String msg = String.format(
+          "Multipart upload failed to complete for bucket '%s' key '%s'. Reason: %s",
+          bucket,
+          key,
+          e.getMessage()
+      );
+      log.error("{}", msg);
+      IOException newIoException = new IOException(msg, e.getCause()) ;
+      // Set original exception's stack trace
+      newIoException.setStackTrace(e.getStackTrace());
+      // Throw the new IOException with updated message and the original cause
+      // (cause is most likely an AmazonClientException)
+      throw newIoException;
     } finally {
       buffer.clear();
       multiPartUpload = null;

--- a/kafka-connect-s3/src/test/java/io/confluent/connect/s3/storage/S3OutputStreamTest.java
+++ b/kafka-connect-s3/src/test/java/io/confluent/connect/s3/storage/S3OutputStreamTest.java
@@ -10,43 +10,53 @@ import com.amazonaws.AmazonServiceException;
 import com.amazonaws.AmazonServiceException.ErrorType;
 import com.amazonaws.services.s3.AmazonS3;
 import io.confluent.connect.s3.S3SinkConnectorTestBase;
-import org.apache.kafka.connect.errors.ConnectException;
 import org.junit.Before;
 import org.junit.Test;
+
+import java.io.IOException;
 
 public class S3OutputStreamTest extends S3SinkConnectorTestBase {
 
   private AmazonS3 s3Mock;
   private S3OutputStream stream;
+  final static String S3_TEST_KEY_NAME = "key";
+  final static String S3_EXCEPTION_MESSAGE = "this is an s3 exception";
+  final String EXPECTED_EXCEPTION_MESSAGE = String.format(
+      "Multipart upload failed to complete for bucket '%s' key '%s'. Reason: %s",
+      S3_TEST_BUCKET_NAME,
+      S3_TEST_KEY_NAME,
+      S3_EXCEPTION_MESSAGE
+  );
+
 
   @Before
   public void before() throws Exception {
     super.setUp();
     s3Mock = mock(AmazonS3.class);
-    stream = new S3OutputStream("key", connectorConfig, s3Mock);
+    stream = new S3OutputStream(S3_TEST_KEY_NAME, connectorConfig, s3Mock);
   }
 
   @Test
   public void testPropagateUnretriableS3Exceptions() {
-    AmazonServiceException e = new AmazonServiceException("this is an s3 exception");
+    AmazonServiceException e = new AmazonServiceException(S3_EXCEPTION_MESSAGE);
     e.setErrorType(ErrorType.Client);
 
     when(s3Mock.initiateMultipartUpload(any())).thenThrow(e);
-    assertThrows("Unable to initiate Multipart Upload.", ConnectException.class, () -> stream.commit());
+    assertThrows(EXPECTED_EXCEPTION_MESSAGE, IOException.class, () -> stream.commit());
   }
 
   @Test
   public void testPropagateRetriableS3Exceptions() {
-    AmazonServiceException e = new AmazonServiceException("this is an s3 exception");
+    AmazonServiceException e = new AmazonServiceException(S3_EXCEPTION_MESSAGE);
     e.setErrorType(ErrorType.Service);
 
     when(s3Mock.initiateMultipartUpload(any())).thenThrow(e);
-    assertThrows("Multipart upload failed to complete.", ConnectException.class, () -> stream.commit());
+    assertThrows(EXPECTED_EXCEPTION_MESSAGE, IOException.class, () -> stream.commit());
   }
 
   @Test
   public void testPropagateOtherRetriableS3Exceptions() {
-    when(s3Mock.initiateMultipartUpload(any())).thenThrow(new AmazonClientException("this is an other s3 exception"));
-    assertThrows("Multipart upload failed to complete.", ConnectException.class, () -> stream.commit());
+    when(s3Mock.initiateMultipartUpload(any())).thenThrow(new AmazonClientException(S3_EXCEPTION_MESSAGE));
+    assertThrows(EXPECTED_EXCEPTION_MESSAGE, IOException.class, () -> stream.commit());
   }
 }

